src/actions/app.js

```diff
@@ -28,10 +30,31 @@ export const toggleEditState = () => ({
   type: TOGGLE_EDITING_STATE,
 });
 
-export const saveAnnotation = (annotationData) => ({
-  type: SAVE_ANNOTATION,
-  payload: annotationData,
-});
+export const saveAnnotation = (annotationData, id = null) => {
+  return (dispatch, getState) => {
+    dispatch({
+      type: SAVE_ANNOTATION,
+      payload: annotationData,
+    });
+    // If the snippet has not been saved, then saving an annotation shouldn't
+    // trigger a POST request to save the snippet (and its annotations)
+    if (!id) {
+      // Return a Promise object to make this action "thenable"
+      return Promise.resolve();
+    }
+    // Get the app state and stringify it
+    const stateString = JSON.stringify(getState().app);
+
+    // TODO: Dispatch action to show the snackbar with a relevant message based on the results
+    // Return the results of the POST request
+    return axios.post(`api/snippets/${id}`, { json: stateString, })
```

Comment:
Okay sure, but now there's coupling between updating state and backend interaction. The http requests are already async thanks to Promises, why not let the thing that's handling the event fire it off?